packages/rpc/json/provider.ts

```diff
@@ -2,43 +2,75 @@ import fetch from "isomorphic-fetch"
 
 import json, { format, replacer } from "~packages/util/json"
 
+import { JsonRpcError } from "./error"
+
+export type JsonRpcResponse<T extends any> = {
+  jsonrpc: "2.0"
+  id: number
+  result?: T
+  error?: {
+    code: number
+    message: string
+    data?: any
+  }
+}
+
 export class JsonRpcProvider {
   public constructor(public readonly rpcUrl: string) {}
 
   public async send<T extends any>(args: {
     method: string
     params?: any[]
   }): Promise<T> {
-    const body = json.stringify(
-      {
+    try {
+      const body = json.stringify(
+        {
+          jsonrpc: "2.0",
+          id: 0,
+          ...args
+        },
+        (k, v) => {
+          if (k === "id") {
+            return v
+          }
+          return replacer.numberToHex(k, v)
+        }
+      )
+      console.log(`[JsonRpcProvider][${args.method}][request] ${body}`)
+
+      const res = await fetch(this.rpcUrl, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json"
+        },
+        body
+      })
+      const payload: JsonRpcResponse<T> = await res.json()
+
+      if (payload.error) {
+        // 200 response but may have issues executing that request
+        throw JsonRpcError.wrap(payload)
+      }
+      console.log(
+        `[JsonRpcProvider][${args.method}][response] ${format(payload)}`
+      )
+      return payload.result
+    } catch (err) {
+      if (err instanceof JsonRpcError) {
+        console.log(`[JsonRpcProvider][${args.method}][error] ${format(err)}`)
+        throw err
+      }
+      console.log(
+        `[JsonRpcProvider][${args.method}][stringify error] ${format(err)}`
```

Comment:
`fetch` might also throw out error when network is broken, could handle it more generally.

```suggestion
        `[JsonRpcProvider][${args.method}][error] ${format(err)}`
```

Reply:
And now the console seems the same as rpc error, maybe we could log first and do other stuffs after it.

Reply:
[log JsonRpcProvider first](https://github.com/pilagod/waallet-extension/pull/179/commits/35138abae8cae9be2e7f652cb92edbcc649e7fee)